### PR TITLE
Generic kubernetes api client

### PR DESCRIPTION
**Pull request by yue9944882**
@brendandburns 

NOTE: i will sanitize the pull l8r (the license, formatting, etc..).. 

it's neat and simple. w/ the help of this generic client, what we only need is the generated model. i just drafted initial version of code and test, it already covers all the api paths on the kubernetes apiserver.

the general idea is to make use of CustomObjectApi as best as we can, this will help us access custom resources in kubernetes in an easier way. appreciated if you feel like shed some light on it

### CODE REVIEW

**Comment by yue9944882** [comment]
note that #86 is solved in this pull, as is shown in the test..

**Comment by yue9944882** [comment]
/hold 

until #742 in

**Comment by brendandburns** [comment]
@yue9944882 is this ready for review?

**Comment by yue9944882** [comment]
@brendandburns let's freeze the pull until the migration to openapi-generator, the work isn't necessary to get into the next release i think.

**Comment by yue9944882** [comment]
/hold cancel

@brendandburns let's move on reviewing this pull. @runzexia feel free to drop your comments too

**Comment by yue9944882** [comment]
@brendandburns bumping for review

File: extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java [hunk 1]
```diff
@@ -0,0 +1,680 @@
+package io.kubernetes.client.extended.generic;
+
+import com.google.common.base.Strings;
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.gson.reflect.TypeToken;
+import io.kubernetes.client.custom.V1Patch;
+import io.kubernetes.client.extended.generic.options.*;
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java]: Please don't use wildcard imports.

File: extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+ */
+public class GenericKubernetesApi<ApiType, ApiListType> {
+
+  // TODO(yue9944882): supports status operations..
+  // TODO(yue9944882): supports generic sub-resoruce operations..
+  // TODO(yue9944882): supports delete-collections..
+
+  private Class<ApiType> apiTypeClass;
+  private Class<ApiListType> apiListTypeClass;
+  private String apiGroup;
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java]: Does it make sense to use the native Kubernetes object here for this? It has these fields.

File: extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+  public GenericKubernetesApi(
+      Class<ApiType> apiTypeClass,
+      Class<ApiListType> apiListTypeClass,
+      String apiGroup,
+      String apiVersion,
+      String resourcePlural,
+      CustomObjectsApi customObjectsApi) {
+    if ("".equals(apiGroup)) {
+      throw new IllegalArgumentException(
+          "core group (\"\") not supported, use CoreV1Api otherwise");
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java]: should this say `use "v1" instead`

File: extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java [hunk 4]
```diff
@@ -0,0 +1,680 @@
...
+    V1ObjectMeta objectMeta;
+    try {
+      objectMeta = Reflect.objectMetadata(object);
+    } catch (ObjectMetaReflectException e) {
+      throw new IllegalArgumentException("fail to extract object metadata");
+    }
+    return executeCall(
+        customObjectsApi.getApiClient(),
+        apiTypeClass,
+        () -> {
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java]: This function looks like it's duplicated with above? If so, pull out with a single implementation.

File: extended/src/main/java/io/kubernetes/client/extended/generic/options/CreateOptions.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package io.kubernetes.client.extended.generic.options;
+
+public class CreateOptions {
+  private Boolean dryRun;
+  private Object
+      fieldManager; // TODO: structured field-manager for better server-side apply integeration
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/generic/options/CreateOptions.java]: why is this on a newline?

File: extended/src/main/java/io/kubernetes/client/extended/generic/options/DeleteOptions.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
+package io.kubernetes.client.extended.generic.options;
+
+import io.kubernetes.client.openapi.models.V1DeleteOptions;
+
+public class DeleteOptions extends V1DeleteOptions {}
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/generic/options/DeleteOptions.java]: what is the purpose of this class?

File: extended/src/main/java/io/kubernetes/client/extended/generic/options/ListOptions.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
+package io.kubernetes.client.extended.generic.options;
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/generic/options/ListOptions.java]: Shouldn't we use the generated class?

File: extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java (hunk 1 above)
Comment (yue9944882) [reply extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java]: fixed

File: extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java (hunk 3 above)
Comment (yue9944882) [reply extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java]: updated to `core group (\"v1\") api not supported, use CoreV1Api instead`, does that sound better?

File: extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java (hunk 4 above)
Comment (yue9944882) [reply extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java]: the called function in the lambda has similar name but different:

replaceNamespacedCustomObjectCall vs replaceClusterCustomObjectCall

File: extended/src/main/java/io/kubernetes/client/extended/generic/options/CreateOptions.java (hunk 5 above)
Comment (yue9944882) [reply extended/src/main/java/io/kubernetes/client/extended/generic/options/CreateOptions.java]: it's the TODO, i moved it to another line

File: extended/src/main/java/io/kubernetes/client/extended/generic/options/DeleteOptions.java (hunk 6 above)
Comment (yue9944882) [reply extended/src/main/java/io/kubernetes/client/extended/generic/options/DeleteOptions.java]: i added a class named `ListOptions` in this package and it's probably good to gather those option class into the same package. the `DeleteOptions` here is just a link to the generated delete option class. wdyt?

File: extended/src/main/java/io/kubernetes/client/extended/generic/options/ListOptions.java (hunk 7 above)
Comment (yue9944882) [reply extended/src/main/java/io/kubernetes/client/extended/generic/options/ListOptions.java]: there is a new class named `ListOption` introduced by this pull. the point of defining option classes here is to provide the developers a better programming interface, which looks making more logical sense comparing to those generated long-list parameters calls. in golang, we have similar option structs defined over the actual API parameters, developer works on that logical options and the SDK translates them to api interface options.. does that make sense to you?

**Comment by brendandburns** [comment]
@yue9944882 fwiw, I'm waiting on responses to the other comments before re-reviewing.

Let me know when it's ready for another pass.

Thanks!

**Comment by yue9944882** [comment]
@brendandburns am i missing anything? i seem to replied the comments on the page. please let me know, thx!

**Comment by brendandburns** [comment]
Was wondering about this comment:

https://github.com/kubernetes-client/java/pull/738#discussion_r363888814

File: extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java (hunk 2 above)
Comment (yue9944882) [reply extended/src/main/java/io/kubernetes/client/extended/generic/GenericKubernetesApi.java]: are you refering to [V1APIGroup](https://github.com/kubernetes-client/java/blob/master/kubernetes/src/main/java/io/kubernetes/client/openapi/models/V1APIGroup.java)? the so-called field "apiGroup" here should be the name of the serving resource api group e.g. `apps`, `networking.k8s.io`, `apiextensions.k8s.io`, etc. while the `V1APIGroup` is used for representing all the fundamental information about the an api group . we need only a tiny bit of that, which is V1APIGroup#name. simply a string value would be sufficient, wdyt?

**Comment by brendandburns** [comment]
I guess it's fine, the plurals and everything else are represented in that object too, afaik.

But this seems like a fine place to start.

/lgtm
/approve

**Pull request merged**